File: Orb_Assistant/user_dock_station/provider-router.js
function trimSlash(value) {
  return String(value || '').trim().replace(/\/+$/, '');
}

function joinUrl(baseUrl, suffix) {
  const base = trimSlash(baseUrl);
  if (!base) return suffix;
  if (base.endsWith(suffix)) return base;
  return `${base}${suffix.startsWith('/') ? '' : '/'}${suffix}`;
}

function extractOpenAIText(payload) {
  return String(
    payload?.choices?.[0]?.message?.content ||
    payload?.choices?.[0]?.text ||
    payload?.output_text ||
    ''
  ).trim();
}

function extractAnthropicText(payload) {
  const blocks = Array.isArray(payload?.content) ? payload.content : [];
  return blocks
    .filter((block) => block?.type === 'text' && block?.text)
    .map((block) => String(block.text))
    .join('\n')
    .trim();
}

function extractGoogleText(payload) {
  const parts = payload?.candidates?.[0]?.content?.parts;
  return Array.isArray(parts)
    ? parts.map((part) => String(part?.text || '')).join('\n').trim()
    : '';
}

async function readFailure(response) {
  const text = await response.text().catch(() => '');
  return text.slice(0, 600) || response.statusText || `HTTP ${response.status}`;
}

async function postJson(url, headers, body, timeoutMs) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...headers
      },
      body: JSON.stringify(body),
      signal: controller.signal
    });
    if (!response.ok) {
      throw new Error(await readFailure(response));
    }
    return response.json();
  } finally {
    clearTimeout(timer);
  }
}

class ProviderRouter {
  constructor({ getRuntimeProfile, getCredential }) {
    if (typeof getRuntimeProfile !== 'function') throw new Error('getRuntimeProfile is required');
    if (typeof getCredential !== 'function') throw new Error('getCredential is required');
    this.getRuntimeProfile = getRuntimeProfile;
    this.getCredential = getCredential;
  }

  orderedSlots(profile) {
    const mode = profile.llm.routingMode;
    const primary = { slot: 'primary', config: profile.llm.primary };
    const fallback = { slot: 'fallback', config: profile.llm.fallback };
    if (mode === 'local_only') {
      return [primary, fallback].filter((entry) => entry.config.provider === 'local_openai_compatible');
    }
    if (mode === 'api_only') {
      return [primary, fallback].filter((entry) => entry.config.provider !== 'local_openai_compatible');
    }
    return profile.llm.fallback.enabled ? [primary, fallback] : [primary];
  }

  async invokeSlot(slot, config, prompt, requestOptions = {}) {
    const provider = config.provider;
    const model = String(config.model || requestOptions.model || '').trim();
    if (!model) throw new Error(`${slot} model is not configured`);
    const apiKey = config.apiKeyStored ? await this.getCredential(slot) : null;
    const timeoutMs = Number(config.timeoutMs || 45000);
    const temperature = Number.isFinite(Number(requestOptions.temperature))
      ? Math.max(0, Math.min(1.5, Number(requestOptions.temperature)))
      : 0.35;
    const maxTokens = Number.isFinite(Number(requestOptions.maxTokens))
      ? Math.max(16, Math.min(2048, Number(requestOptions.maxTokens)))
      : 180;

    if (provider === 'anthropic') {
      if (!apiKey) throw new Error('Anthropic API key is not configured');
      const baseUrl = trimSlash(config.baseUrl) || 'https://api.anthropic.com/v1';
      const payload = await postJson(
        joinUrl(baseUrl, '/messages'),
        {
          'x-api-key': apiKey,
          'anthropic-version': '2023-06-01'
        },
        {
          model,
          max_tokens: maxTokens,
          temperature,
          messages: [{ role: 'user', content: prompt }]
        },
        timeoutMs
      );
      const text = extractAnthropicText(payload);
      if (!text) throw new Error('Anthropic returned no text');
      return { text, provider, model };
    }

    if (provider === 'google') {
      if (!apiKey) throw new Error('Google API key is not configured');
      const baseUrl = trimSlash(config.baseUrl) || 'https://generativelanguage.googleapis.com/v1beta';
      const endpoint = `${joinUrl(baseUrl, `/models/${encodeURIComponent(model)}:generateContent`)}?key=${encodeURIComponent(apiKey)}`;
      const payload = await postJson(
        endpoint,
        {},
        {
          contents: [{ role: 'user', parts: [{ text: prompt }] }],
          generationConfig: {
            temperature,
            maxOutputTokens: maxTokens
          }
        },
        timeoutMs
      );
      const text = extractGoogleText(payload);
      if (!text) throw new Error('Google returned no text');
      return { text, provider, model };
    }

    const baseUrl = trimSlash(config.baseUrl) || (provider === 'openai' ? 'https://api.openai.com/v1' : 'http://127.0.0.1:40343/v1');
    if (provider !== 'local_openai_compatible' && !apiKey) {
      throw new Error(`${provider} API key is not configured`);
    }
    const payload = await postJson(
      joinUrl(baseUrl, '/chat/completions'),
      apiKey ? { Authorization: `Bearer ${apiKey}` } : {},
      {
        model,
        messages: [{ role: 'user', content: prompt }],
        temperature,
        max_tokens: maxTokens,
        stream: false
      },
      timeoutMs
    );
    const text = extractOpenAIText(payload);
    if (!text) throw new Error(`${provider} returned no text`);
    return { text, provider, model };
  }

  async generate({ prompt, model, options = {} }) {
    const cleanPrompt = String(prompt || '').trim();
    if (!cleanPrompt) throw new Error('Prompt is required');
    const profile = this.getRuntimeProfile();
    const attempts = [];
    for (const entry of this.orderedSlots(profile)) {
      if (entry.slot === 'fallback' && !profile.llm.fallback.enabled) continue;
      try {
        const result = await this.invokeSlot(entry.slot, entry.config, cleanPrompt, {
          model,
          temperature: options.temperature,
          maxTokens: options.num_predict
        });
        return {
          ...result,
          slot: entry.slot,
          profileRevision: profile.revision
        };
      } catch (error) {
        attempts.push({
          slot: entry.slot,
          provider: entry.config.provider,
          error: String(error?.message || error).slice(0, 500)
        });
      }
    }
    const failure = new Error('No configured model provider completed the request');
    failure.attempts = attempts;
    throw failure;
  }
}

module.exports = {
  ProviderRouter,
  extractOpenAIText,
  extractAnthropicText,
  extractGoogleText
};

File: backend/app/orb/provider_router.py
from __future__ import annotations

import os
from typing import Any, Dict, Optional

import httpx


def _join_url(base: str, suffix: str) -> str:
    clean = (base or "").strip().rstrip("/")
    return clean if clean.endswith(suffix) else f"{clean}{suffix}"


def _credential(configuration: Dict[str, Any]) -> Optional[str]:
    variable = str(configuration.get("api_key_env") or "").strip()
    return os.environ.get(variable) if variable else None


def _extract_openai(payload: Dict[str, Any]) -> str:
    choices = payload.get("choices") or []
    if choices:
        return str((choices[0].get("message") or {}).get("content") or choices[0].get("text") or "").strip()
    return str(payload.get("output_text") or "").strip()


def _extract_anthropic(payload: Dict[str, Any]) -> str:
    return "\n".join(
        str(block.get("text") or "")
        for block in payload.get("content") or []
        if block.get("type") == "text"
    ).strip()


def _extract_google(payload: Dict[str, Any]) -> str:
    candidates = payload.get("candidates") or []
    if not candidates:
        return ""
    return "\n".join(
        str(part.get("text") or "")
        for part in (candidates[0].get("content") or {}).get("parts") or []
    ).strip()


async def invoke_provider(
    configuration: Dict[str, Any],
    *,
    prompt: str,
    system_instruction: str,
    client: Optional[httpx.AsyncClient] = None,
) -> Dict[str, Any]:
    provider = str(configuration.get("provider") or "runtime_default")
    model = str(configuration.get("model") or "").strip()
    timeout = min(120.0, max(5.0, float(configuration.get("timeout_seconds") or 45.0)))
    result: Dict[str, Any] = {"success": False, "provider": provider, "model": model, "text": None, "error": None}
    if provider == "runtime_default":
        return {**result, "error": "provider_not_requested"}
    if not model:
        return {**result, "error": "provider_model_missing"}

    api_key = _credential(configuration)
    owns_client = client is None
    http = client or httpx.AsyncClient(timeout=timeout)
    try:
        if provider == "anthropic_api":
            if not api_key:
                return {**result, "error": "provider_credential_missing"}
            base = str(configuration.get("base_url") or "https://api.anthropic.com/v1")
            response = await http.post(
                _join_url(base, "/messages"),
                headers={"x-api-key": api_key, "anthropic-version": "2023-06-01"},
                json={"model": model, "system": system_instruction, "max_tokens": 220, "messages": [{"role": "user", "content": prompt}]},
            )
            response.raise_for_status()
            text = _extract_anthropic(response.json())
        elif provider == "google_api":
            if not api_key:
                return {**result, "error": "provider_credential_missing"}
            base = str(configuration.get("base_url") or "https://generativelanguage.googleapis.com/v1beta")
            response = await http.post(
                f"{_join_url(base, f'/models/{model}:generateContent')}?key={api_key}",
                json={
                    "systemInstruction": {"parts": [{"text": system_instruction}]},
                    "contents": [{"role": "user", "parts": [{"text": prompt}]}],
                    "generationConfig": {"temperature": 0.35, "maxOutputTokens": 220},
                },
            )
            response.raise_for_status()
            text = _extract_google(response.json())
        else:
            if provider == "openai_api":
                base = str(configuration.get("base_url") or "https://api.openai.com/v1")
                if not api_key:
                    return {**result, "error": "provider_credential_missing"}
            elif provider in {"openai_compatible", "local_openai_compatible"}:
                base = str(configuration.get("base_url") or "").strip()
                if not base:
                    return {**result, "error": "provider_endpoint_missing"}
            else:
                return {**result, "error": "provider_unsupported"}
            response = await http.post(
                _join_url(base, "/chat/completions"),
                headers={"Authorization": f"Bearer {api_key}"} if api_key else {},
                json={
                    "model": model,
                    "messages": [{"role": "system", "content": system_instruction}, {"role": "user", "content": prompt}],
                    "temperature": 0.35,
                    "max_tokens": 220,
                    "stream": False,
                },
            )
            response.raise_for_status()
            text = _extract_openai(response.json())
        if not text:
            return {**result, "error": "provider_empty_response"}
        return {**result, "success": True, "text": text}
    except httpx.HTTPStatusError as exc:
        return {**result, "error": f"provider_request_failed:http_{exc.response.status_code}"}
    except httpx.HTTPError as exc:
        return {**result, "error": f"provider_request_failed:{exc.__class__.__name__}"}
    except ValueError:
        return {**result, "error": "provider_request_failed:invalid_response"}
    finally:
        if owns_client:
            await http.aclose()
